### ISSUE TITLE
Ensure all failed indirect peer connections send CannotConnect if neither a direct nor indirect connection was established

### ISSUE THREAD

**Issue by jpdillingham** [jpdillingham/Soulseek.NET#514]
This doesn't appear to be happening for parent candidates during the adoption process.  Review other parts of the code to make sure it is being sent in all cases; messaging, transfer and distributed connections to other peers, parents and children.

I was initially less concerned about sending these, thinking the message had been deprecated, but Soulseek Qt sends them when failing to connect to potential parents which may impact the selection process.

**Comment by mathiascode** [comment]
I realized this message is still in use at a later point, but haven't gotten around to re-implementing it yet. It seems like SoulseekQt sends it after trying to respond to an indirect connection request for 60 seconds, as opposed to sending it on first failure, which was old Nicotine+ behavior that caused issues when indirect requests arrived before direct ones.

**Comment by jpdillingham** [comment]
Yeah, I took them out after we traded some messages about it for the same reasons; it was creating issues because I was sending it before the client on the other end was done trying to connect.  I think my plan will be to add a delay two times longer than the configured connection timeout, to be sure the message would be sent clear of any concurrent direct connection.

Do you have a feel for whether this message is sent if the indirect connection failed, but a direct connection succeeded?  (I sure hope not because that would be difficult to implement)

**Comment by jpdillingham** [comment]
I confirmed that Qt only sends this when it fails to establish both direct and indirect connections.  If a direct connection is established, it appears to just silently discard the indirect attempt.  NS seems to send it immediately and regardless of the status of the direct connection.

My handling of direct and indirect incoming connections is segregated enough that this would be a non-trivial thing to implement so this is likely not going to be a priority since I'm not 100% sure there's any benefit to anyone.

**Comment by mathiascode** [comment]
> I confirmed that Qt only sends this when it fails to establish both direct and indirect connections.

Makes sense. Afaik, the only purpose of the message is to inform the other peer that any form of connection method has failed (at least in Nicotine+, although the message is useless for us, since we have a local timeout for indirect connection requests).

I'm not a fan of sending this message again, due to the concern of clogging up connection slots for popular searches, in cases where both you and the sender have a closed port. Even if an indirect connection request is only responded to once, keeping thousands of connection objects around for max ~60 seconds would degrade the performance.

**Comment by jpdillingham** [comment]
Agreed, my two options would be to kick off a `Task` with a 60 second delay or to stuff them into a dictionary, which would then require a timer to sweep it periodically.  I want to avoid thrashing the thread pool and am already not a huge fan of the couple of timers I've had to implement so far.

When I opened this I thought that maybe the server would use this feedback to determine which parents are sent in `NetInfo` but watching the diagnostic window of Qt reveals that it happily sends the same parents over and over.

I see only downsides in this.
